Firestore/Example/Tests/Integration/FSTTransactionTests.mm

```diff
@@ -20,184 +20,346 @@
+          stage(transaction, self->_docRef);
+        }
         return @YES;
       }
       completion:^(id _Nullable result, NSError *_Nullable error) {
-        XCTAssertEqualObjects(@YES, result);
-        XCTAssertNil(error);
         [expectation fulfill];
+        NSString *message =
+            [NSString stringWithFormat:@"Expected the sequence (%@), to fail, but it didn't.",
+                                       [self listStages]];
+        [self->_testCase assertErrorWithMessage:message error:error];
       }];
-  [self awaitExpectations];
 
-  snapshot = [self readDocumentForRef:doc];
+  [_testCase awaitExpectations];
+}
+
+- (void)cleanupTester {
+  _stages = [NSArray array];
+  // Set the docRef to something else to lose the original reference.
+  _docRef = [[self->_db collectionWithPath:@"reset"] documentWithAutoID];
+}
+
+- (NSString *)listStages {
+  NSMutableArray<NSString *> *seqList = [NSMutableArray array];
+  for (TransactionStage stage in _stages) {
+    if (stage == delete1) {
+      [seqList addObject:@"delete"];
+    } else if (stage == update1 || stage == update2) {
+      [seqList addObject:@"update"];
+    } else if (stage == set1 || stage == set2) {
+      [seqList addObject:@"set"];
+    } else if (stage == get) {
+      [seqList addObject:@"get"];
+    }
+  }
+  return [seqList description];
+}
+
+@end
+
+@implementation FSTTransactionTests
+- (void)assertExistsWithSnapshot:(FIRDocumentSnapshot *)snapshot error:(NSError *)error {
+  XCTAssertNil(error);
+  XCTAssertTrue(snapshot.exists);
+}
+
+- (void)assertDoesNotExistWithSnapshot:(FIRDocumentSnapshot *)snapshot error:(NSError *)error {
+  XCTAssertNil(error);
   XCTAssertFalse(snapshot.exists);
 }
 
-- (void)testGetNonexistentDocumentThenCreate {
-  FIRFirestore *firestore = [self firestore];
-  FIRDocumentReference *doc = [[firestore collectionWithPath:@"towns"] documentWithAutoID];
+- (void)assertNilErrorWithMessage:(NSString *)message error:(NSError *)error {
+  XCTAssertNil(error, @"%@", message);
+}
 
-  XCTestExpectation *expectation = [self expectationWithDescription:@"transaction"];
-  [firestore
-      runTransactionWithBlock:^id _Nullable(FIRTransaction *transaction, NSError **error) {
-        FIRDocumentSnapshot *snapshot = [transaction getDocument:doc error:error];
-        XCTAssertNil(*error);
-        XCTAssertNotNil(snapshot);
-        XCTAssertFalse(snapshot.exists);
-        [transaction setData:@{@"foo" : @"bar"} forDocument:doc];
-        return @YES;
-      }
-      completion:^(id _Nullable result, NSError *_Nullable error) {
-        XCTAssertEqualObjects(@YES, result);
-        XCTAssertNil(error);
-        [expectation fulfill];
-      }];
-  [self awaitExpectations];
+- (void)assertErrorWithMessage:(NSString *)message error:(NSError *)error {
+  XCTAssertNotNil(error, @"%@", message);
+}
 
-  FIRDocumentSnapshot *snapshot = [self readDocumentForRef:doc];
+- (void)assertEqualsObjectWithSnapshot:(FIRDocumentSnapshot *)snapshot
```

Comment:
Try to phrase this such that the label and its parameters reads approximately like English text.

How about `assertSnapshot:equalsObject:error:`?